### PR TITLE
feat(pop-api): charge proper weights in chain-extension

### PR DESCRIPTION
**Pull request by peterwht**
Prerequisite: Merge https://github.com/r0gue-io/pop-node/pull/22 first

### CODE REVIEW

File: runtime/src/extensions.rs [hunk 1]
```diff
@@ -87,77 +89,95 @@ where
...
+	const LOG_PREFIX: &str = " read_state |";
 
 	let mut env = env.buf_in_buf_out();
 
-	// TODO: Substitute len u32 with pop_api::src::impls::pop_network::StringLimit.
-	// Move StringLimit to pop_api_primitives first.
-	let len: u32 = env.in_len();
-	let key: ParachainSystemKeys = env.read_as_unbounded(len)?;
+	// To be conservative, we charge the weight for reading the input bytes.
+	// However, charge weight is not strictly necessary here as reading a fixed value's weight is included in the contract call.
```
Comment (evilrobot-01) [inline runtime/src/extensions.rs:150]: Not sure this is entirely accurate. `read_as` does indeed use `read_sandbox_memory_as`, but the description of `read_as` states more about the weight of a chain extension, where the function is located, whereas `read_sandbox_memory_as` is a general purpose host function. I think it depends on the context, so would still be cautious for now.

File: runtime/src/extensions.rs [hunk 2]
```diff
@@ -87,77 +89,95 @@ where
...
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
 		},
 		Err(err) => {
-			log::debug!(target:LOG_TARGET, "dispatch failed: error: {:?}", err.error);
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", LOG_PREFIX, err.error);
 			return Err(err.error);
 		},
 	}
 	Ok(())
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply runtime/src/extensions.rs]: Pushed a commit to sort this.

**Review by evilrobot-01 — APPROVED** [review]
A few comments worth discussing, but apart from that LGTM.

File: runtime/src/extensions.rs (hunk 1 above)
Comment (Daanvdplas) [reply runtime/src/extensions.rs:150]: What is meant with a fixed value's weight?

**Review by Daanvdplas — APPROVED** [review]

File: runtime/src/extensions.rs [hunk 3]
```diff
@@ -87,77 +89,95 @@ where
...
-	let _ = env.charge_weight(base_weight.saturating_add(dispatch_weight))?;
+	// read the input as RuntimeCall
+	let call: <T as SysConfig>::RuntimeCall = env.read_as_unbounded(len)?;
 
-	log::debug!(target:LOG_TARGET, " dispatch inputted RuntimeCall: {:?}", call);
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+
+	log::debug!(target:LOG_TARGET, "{} inputted RuntimeCall: {:?}", LOG_PREFIX, call);
 
 	let sender = env.ext().caller();
```
[the remark this answers is not in the document]
Comment (peterwht) [reply runtime/src/extensions.rs:119]: Yes, _only_ contracts can call chain extensions

**Pull request merged**